### DISCUSSION TITLE
dispatch a createAsyncThunk when a reducer si called

### DISCUSSION THREAD

**Discussion by Piero87** [reduxjs/redux-toolkit#1765] [Q&A] [answered]
Hello,
I want refresh the content of the page when a state in the slice is changed, I add a sample code just to explain me better:

export const getFeedbacks = createAsyncThunk('feedbacks/get', async (arg, { getState }) => {
...api....
return data[...];
});

const feedbacksSlice = createSlice({
  name: 'feedbacks',
  initialState: {
     filter: 0,
     data: []
  },
  reducers: {
    changeFilter: (state, action) => {
      state.filter = action.payload;
      // HERE>>>>> I want call getFeedbacks
    },
    extraReducers(builder) {
    builder.addCase.......
    .....
  },

so when the `changeFilter` is called I want to call the `getFeedbacks` to refresh data, where I have to make this call? inside the reducer? inside the react component listening for a change of `filter` inside an useEffect?

**Comment by markerikson** [comment, marked as answer]
No, **you can _never_ make any kind of additional call like that in a reducer**.

However, you _could_ try using the new experimental "action listener middleware" we're looking to add to RTK, currently available as a standalone package:

https://github.com/reduxjs/redux-toolkit/discussions/1648

**Reply by Piero87** [reply]
@markerikson thank you very much for the answer, that middleware seems to be very interesting.

otherwise with out it at the moment what would be the right way to do it?

**Reply by Piero87** [reply]
@markerikson I'm reading the documentation of the "action listener middleware" you are working on, and it's very very interesting, I want try to use it, I have some questions about it:

1. Can the listener be added in the slice file interested of this listener? if yes how can I add it dynamically? I read in the documentation that can be added with a dispatch, but I can't understand how
2. About the question above, it's the right way (if it's possible for the answer of the question 1) to addlistener when the "feature/component/page" is mounted and remove it when it unmounted?

**Reply by markerikson** [reply]
The middleware itself needs to be added to the store at setup time (ie, passed as part of `configureStore()`).

Listeners can be added at any time, either by importing the `middleware` instance directly and calling `middleware.addListener()`, or by doing `dispatch(addListenerAction())`.  Either of those returns an `unsubscribe` function.

So yes, you could definitely do something like:

```js
const dispatch = useDispatch();

useEffect(() => {
  const unsubscribe = dispatch(addListenerAction(
    actionCreator: incrementByAmount,
    listener: someListenerFunctionHere
  ));
  
  return unsubscribe;
}, [])
```

and it should get added on mount and removed on unmount.

This pattern is not necessarily a _great_ idea, as components should ideally be paying more attention to the store state instead of specific actions, but it is valid.

**Reply by Piero87** [reply]
@markerikson Thank you very much now it's becoming clearer to me, I miss the `addListenerAction` or the importing the middleware what you mean? can you please make me an example of the addListenerAction and importing the middleware?

I have one another question about this thing you sad:

> This pattern is not necessarily a great idea, as components should ideally be paying more attention to the store state instead of specific actions, but it is valid.

What do you mean? where did you add this listeners you will have in the application?

**Reply by markerikson** [reply]
The main examples for now would be looking at [the test file for the middleware package](https://github.com/reduxjs/redux-toolkit/blob/69d6965c56ad20078d43a297509b2f380e73592b/packages/action-listener-middleware/src/tests/listenerMiddleware.test.ts).

"Importing the middleware" would likely be:

```js
// src/app/listenerMiddleware.js

import { createActionListenerMiddleware } from '@rtk-incubator/action-listener-middleware';

export const listenerMiddleware = createActionListenerMiddleware();

// someOtherFile.js
import { listenerMiddleware } from 'app/listenerMiddleware'

listenerMiddleware.addListener(/* etc */)
```

The `addListenerAction` action creator is exported from the package.


Per your other question: middleware are global and added to the store once at setup.  As such, we also mostly expect that most listeners are static and added once to the listener middleware as part of the app setup process.

You _can_ add and remove listeners at runtime, from anywhere - we specifically included `addListenerAction` to make that possible.  But as @phryneas has pointed out, Redux is not entirely meant to be used as a pure "event bus", and if components start snooping on actions all the time, that's sort of what you're doing.

So, can you? Yes.  _Should_ you? Maybe :)

**Reply by Piero87** [reply]
@markerikson Thank you very much for the great explanation, so you suggest to add the listener (or most of the listeners) and don't remove them but be part of the setup like the registration of a reducer in the configure store.

Thank you Mark!

**Reply by markerikson** [reply]
Yeah, "add some listeners once on app setup" is the expected primary use case here, but "dynamic add/remove" _is_ something the middleware explicitly supports because it's potentially useful.  (in fact, the `condition` option itself is just adding another listener at runtime.)

**Reply by Piero87** [reply]
Yes all clear, at the moment my main target is to write this listener inside the slice where the action of which they are listening reside, to do this I can use the import of the middleware like you suggest and the addListener function, to have all logic of the slice in one place (at the moment I will not unsubscribe those listener)

**Comment by Piero87** [comment]
@markerikson I'm trying to use it but I have this error:

`action-listener-middleware-error TypeError: i.listener is not a function`

I have make the simpler inclusion:

```
import { configureStore } from '@reduxjs/toolkit';
import { createActionListenerMiddleware } from '@rtk-incubator/action-listener-middleware';


/**
 * Reducers
 */
import feedbacksReducer, {changeFilter} from 'features/feedbacks/feedbacksSlice';

const listenerMiddleware = createActionListenerMiddleware();

listenerMiddleware.addListener(changeFilter, (action, listenerApi) => {
  // Run whatever additional side-effect-y logic you want here
  console.log(action);
});


export default configureStore({
  reducer: {
    feedbacks: feedbacksReducer
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(listenerMiddleware),
});
```

what am I doing wrong?

**Reply by markerikson** [reply]
As of v0.3.0, the `addListener` function takes an object with multiple options, like:

`middleware.addListener({actionCreator: incrementByAmount: listener: (action, listenerApi) => {})`

Possible options for the way to trigger the listener are `type`, `actionCreator`, `matcher`, or `predicate`.

Uh... whoops. Looks like I updated the code and forgot to update the docs in the README or in that discussion issue :)

See that test file I linked above for examples of the arguments.  I'll try to do some updates to the docs later this weekend, but won't get to it tonight.

**Reply by Piero87** [reply]
@markerikson Ok thank you, I changed to this:

```
listenerMiddleware.addListener(({actionCreator: changeFilter, listener}) => {
  // Run whatever additional side-effect-y logic you want here
  console.log(listener);
});
```

but now I get this error:

` TypeError: i.predicate is not a function`

**Reply by Piero87** [reply]
Also this, same error:

```
listenerMiddleware.addListener({
  actionCreater: changeFilter,
  listener: (action, listenerApi) => {
    console.log(action);
  },
});
```

**Reply by markerikson** [reply]
Hmm. We're getting off topic here, but can you put up a CodeSandbox that shows this happening?

**Reply by Piero87** [reply]
Sorry to go offtopic, yes I made a codesandbox for you:

https://codesandbox.io/s/sharp-easley-1ucn9

if you press the test button you will see the error that I see

thanks for the help

**Reply by Piero87** [reply]
@markerikson Please tell me if I have to open a new bug discussion, thank you again!

**Reply by markerikson** [reply]
Aha! You have a typo in your code. You wrote `actionCreater`, but it's `actionCreator` (with an 'o'):

```js
listenerMiddleware.addListener({
  actionCreater: changeFilter, // ERROR should be actionCreator
  listener: (action, listenerApi) => {
    console.log(action);
  }
});
```

So, the middleware code couldn't find the right field and crashed later.

**Reply by markerikson** [reply]
(note that TypeScript would have caught this :)  )

**Reply by Piero87** [reply]
@markerikson Thank you very much!!! Yes I'm seriously considering switching to typescript
